File: hw/palomo/u3/hw24CompositePattern/Composite.ts/src/model/Supervisor.ts
import { Employee } from "./Employee";

export class Supervisor extends Employee {
    protected subordinates: Employee[] = [];

    constructor(name: string) {
        super(name);
    }

    protected getRole(): string {
        return "Supervisor";
    }

    addEmployee(employee: Employee): void {
        this.subordinates.push(employee);
    }

    removeEmployee(employee: Employee): void {
        this.subordinates = this.subordinates.filter((e) => e !== employee);
    }

    stateName(): void {
        super.stateName();
        for (const employee of this.subordinates) {
            employee.stateName();
        }
    }
}

File: hw/palomo/u3/hw24CompositePattern/Composite/src/main/java/ec/edu/espe/composite/model/Supervisor.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

package ec.edu.espe.composite.model;

import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author Cristian Palomo, Error 404, @ESPE
 */
public class Supervisor extends Employee{
    
    protected final List<Employee> subordinates = new ArrayList<>();
    
    public Supervisor(String name){
        super(name);
    }
    
    @Override
    protected String getRole() {
        return "Supervisor";
    }

    public void addEmployee(Employee employee) {
        subordinates.add(employee);
    }

    public void removeEmployee(Employee employee) {
        subordinates.remove(employee);
    }

    @Override
    public void stateName() {
        super.stateName();
        for (Employee employee : subordinates) {
            employee.stateName();
        }
    }
}
